planning/behavior_path_planner/src/scene_module/avoidance_by_lc/module.cpp

```diff
@@ -0,0 +1,1380 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "behavior_path_planner/scene_module/avoidance_by_lc/module.hpp"
+
+#include "behavior_path_planner/path_utilities.hpp"
+#include "behavior_path_planner/scene_module/scene_module_interface.hpp"
+#include "behavior_path_planner/scene_module/scene_module_visitor.hpp"
+#include "behavior_path_planner/turn_signal_decider.hpp"
+#include "behavior_path_planner/util/avoidance/util.hpp"
+#include "behavior_path_planner/util/lane_change/util.hpp"
+#include "behavior_path_planner/utilities.hpp"
+
+#include <lanelet2_extension/utility/message_conversion.hpp>
+#include <lanelet2_extension/utility/utilities.hpp>
+#include <tier4_autoware_utils/tier4_autoware_utils.hpp>
+
+#include "tier4_planning_msgs/msg/detail/lane_change_debug_msg_array__struct.hpp"
+#include <autoware_auto_perception_msgs/msg/object_classification.hpp>
+#include <autoware_auto_vehicle_msgs/msg/turn_indicators_command.hpp>
+
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace behavior_path_planner
+{
+
+using autoware_auto_perception_msgs::msg::ObjectClassification;
+using motion_utils::calcSignedArcLength;
+using motion_utils::findNearestIndex;
+using motion_utils::findNearestSegmentIndex;
+using tier4_autoware_utils::calcDistance2d;
+using tier4_autoware_utils::calcLateralDeviation;
+
+AvoidanceByLCModule::AvoidanceByLCModule(
+  const std::string & name, rclcpp::Node & node,
+  const std::shared_ptr<AvoidanceByLCParameters> parameters,
+  const std::unordered_map<std::string, std::shared_ptr<RTCInterface>> & rtc_interface_ptr_map)
+: SceneModuleInterface{name, node, rtc_interface_ptr_map}, parameters_{parameters}
+{
+  steering_factor_interface_ptr_ =
+    std::make_unique<SteeringFactorInterface>(&node, "avoidance_by_lane_change");
+}
+
+void AvoidanceByLCModule::processOnEntry()
+{
+#ifndef USE_OLD_ARCHITECTURE
+  waitApproval();
+#endif
+  current_lane_change_state_ = LaneChangeStates::Normal;
+  updateLaneChangeStatus();
+}
+
+void AvoidanceByLCModule::processOnExit()
+{
+  resetParameters();
+}
+
+bool AvoidanceByLCModule::isExecutionRequested() const
+{
+  if (current_state_ == ModuleStatus::RUNNING) {
+    return true;
+  }
+
+#ifdef USE_OLD_ARCHITECTURE
+  const auto current_lanes = util::getCurrentLanes(planner_data_);
+#else
+  const auto current_lanes =
+    util::getCurrentLanesFromPath(*getPreviousModuleOutput().reference_path, planner_data_);
+#endif
+  const auto lane_change_lanes = getLaneChangeLanes(current_lanes, lane_change_lane_length_);
+
+  LaneChangePath selected_path;
+  const auto [found_valid_path, found_safe_path] =
+    getSafePath(lane_change_lanes, check_distance_, selected_path);
+
+  if (!found_safe_path) {
+    return false;
```

Comment:
Because lane change checked for valid path and not safe path during `isExecutionRequested`, we don't concern with the path is safe or not.

How about in Avoidance by Lane change, is it by intention to choose `found_safe_path`?
(i guess it makes sense to avoid only when safe path is found)

Reply:
oh, it is also my mistake.

Reply:
I fixed in https://github.com/autowarefoundation/autoware.universe/pull/3125/commits/fb573208e1ff9dbe77726f59da5cf82996c89fc7